File: packages/pvm-assemblyscript/assembly/host-functions/accumulate/query.ts
import { RESULT_CODE_PANIC } from '../../config'
import { getRequestValue, decodeRequestTimeslots } from '../../codec'
import {
  ACCUMULATE_ERROR_NONE,
  AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
  HostFunctionResult,
} from './base'

/**
 * QUERY accumulation host function (Ω_Q)
 *
 * Queries preimage request status
 *
 * Gray Paper Specification:
 * - Function ID: 22 (query)
 * - Gas Cost: 10
 * - Parameters: registers[7-8] = o, z
 *   - o: hash offset in memory (32 bytes)
 *   - z: size of the preimage
 * - Returns: registers[7-8] = status and additional data
 *
 * Gray Paper Logic:
 * 1. Read hash from memory (32 bytes)
 * 2. Look up request status in current service's requests
 * 3. Return encoded status:
 *    - NONE: request doesn't exist
 *    - 0: request exists but is empty []
 *    - 1 + 2^32 * x: request has one entry [x]
 *    - 2 + 2^32 * x, y: request has two entries [x, y]
 *    - 3 + 2^32 * x, y + 2^32 * z: request has three entries [x, y, z]
 */
export class QueryHostFunction extends BaseAccumulateHostFunction {
  functionId: u64 = u64(22) // QUERY function ID (Gray Paper: query = 22)
  name: string = 'query'
  gasCost: u64 = u64(10)

  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const registers = context.registers
    const ram = context.ram
    const implications = context.implications
    // Gray Paper line 204: Ω_F receives H_timeslot (block header's timeslot)
    // This is the current block's timeslot passed from the Accumulate invocation

    // Extract parameters from registers
    const preimageOffset = u64(registers[7])
    const preimageLength = u64(registers[8])

    // Read hash from memory (32 bytes)
    // Gray Paper: h = memory[o:32] - always reads 32 bytes (hash size)
    const readResult_hash = ram.readOctets(u32(preimageOffset), 32)
    if (readResult_hash.faultAddress !== 0) {
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }
    if (readResult_hash.data === null) {
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }
    const hashData = readResult_hash.data!

    // Get the current implications context
    const imX = implications.regular

    // Get current service account
    const accountEntry = this.findAccountEntry(imX.state.accounts, imX.id)
    if (accountEntry === null) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_NONE)
      registers[8] = u64(0)
      return new HostFunctionResult(255) // continue execution
    }
    const serviceAccount = accountEntry.account

    // Look up request using rawCshKeyvals helper
    const requestValue = getRequestValue(serviceAccount, u32(imX.id), hashData, preimageLength)

    if (requestValue === null) {
      // Request doesn't exist
      this.setAccumulateError(registers, ACCUMULATE_ERROR_NONE)
      registers[8] = u64(0)
      return new HostFunctionResult(255) // continue execution
    }

    // Decode the request timeslots
    const timeslots = decodeRequestTimeslots(requestValue)
    if (timeslots === null) {
      // Invalid request value
      this.setAccumulateError(registers, ACCUMULATE_ERROR_NONE)
      registers[8] = u64(0)
      return new HostFunctionResult(255) // continue execution
    }

    // Return encoded status based on request length
    const TWO_TO_32: u64 = u64(4294967296) // 2^32
    if (timeslots.length === 0) {
      // Empty request []
      registers[7] = u64(0)
      registers[8] = u64(0)
    } else if (timeslots.length === 1) {
      // Single entry [x]
      const x = u64(timeslots[0])
      registers[7] = u64(1) + TWO_TO_32 * x
      registers[8] = u64(0)
    } else if (timeslots.length === 2) {
      // Two entries [x, y]
      const x = u64(timeslots[0])
      const y = u64(timeslots[1])
      registers[7] = u64(2) + TWO_TO_32 * x
      registers[8] = y
    } else if (timeslots.length === 3) {
      // Three entries [x, y, z]
      const x = u64(timeslots[0])
      const y = u64(timeslots[1])
      const z = u64(timeslots[2])
      registers[7] = u64(3) + TWO_TO_32 * x
      registers[8] = y + TWO_TO_32 * z
    } else {
      // Invalid request state
      this.setAccumulateError(registers, ACCUMULATE_ERROR_NONE)
      registers[8] = u64(0)
      return new HostFunctionResult(255) // continue execution
    }

    return new HostFunctionResult(255) // continue execution
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/query.rs
//! QUERY accumulation host function (Ω_Q). Gray Paper: function ID 22.
//! r7 = hash offset (o), r8 = preimage size (z). 1:1 with AS query.ts.
//! Read 32-byte hash; look up request; return encoded status: NONE / 0 / 1+2^32*x / 2+2^32*x,y / 3+2^32*x,y+2^32*z.

use crate::config::FUNC_QUERY;
use crate::host_functions::accumulate::base::{self, codes};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};
use crate::codec::{decode_request_timeslots, get_request_value};

const HASH_LEN: u32 = 32;
const TWO_TO_32: u64 = 4_294_967_296; // 2^32

pub struct QueryHostFunction;

impl HostFunction for QueryHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_QUERY
    }
    fn name(&self) -> &'static str {
        "query"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let preimage_offset = context.registers[7] as u32;
        let preimage_length = context.registers[8];

        // Read hash from memory (32 bytes). Gray Paper: panic when read fails; r7 unchanged.
        let read_result = context.ram.read_octets(preimage_offset, HASH_LEN);
        if read_result.fault_address != 0 || read_result.data.is_none() {
            return HostFunctionResult::panic();
        }
        let hash_data = read_result.data.unwrap();
        if hash_data.len() != HASH_LEN as usize {
            return HostFunctionResult::panic();
        }

        // Current service account (imX). None → NONE, r8=0.
        let service_account = match &context.service_account {
            Some(acc) => acc,
            None => {
                base::set_accumulate_error(context.registers, codes::NONE);
                context.registers[8] = 0;
                return HostFunctionResult::continue_execution();
            }
        };
        let service_id = match context.service_id {
            Some(id) => id as u32,
            None => {
                base::set_accumulate_error(context.registers, codes::NONE);
                context.registers[8] = 0;
                return HostFunctionResult::continue_execution();
            }
        };

        let request_value = get_request_value(service_account, service_id, &hash_data, preimage_length);

        let timeslots = match request_value {
            None => {
                base::set_accumulate_error(context.registers, codes::NONE);
                context.registers[8] = 0;
                return HostFunctionResult::continue_execution();
            }
            Some(ref value) => match decode_request_timeslots(value) {
                Some(t) => t,
                None => {
                    base::set_accumulate_error(context.registers, codes::NONE);
                    context.registers[8] = 0;
                    return HostFunctionResult::continue_execution();
                }
            },
        };

        match timeslots.len() {
            0 => {
                context.registers[7] = 0;
                context.registers[8] = 0;
            }
            1 => {
                let x = u64::from(timeslots[0]);
                context.registers[7] = 1 + TWO_TO_32 * x;
                context.registers[8] = 0;
            }
            2 => {
                let x = u64::from(timeslots[0]);
                let y = u64::from(timeslots[1]);
                context.registers[7] = 2 + TWO_TO_32 * x;
                context.registers[8] = y;
            }
            3 => {
                let x = u64::from(timeslots[0]);
                let y = u64::from(timeslots[1]);
                let z = u64::from(timeslots[2]);
                context.registers[7] = 3 + TWO_TO_32 * x;
                context.registers[8] = y + TWO_TO_32 * z;
            }
            _ => {
                base::set_accumulate_error(context.registers, codes::NONE);
                context.registers[8] = 0;
            }
        }

        HostFunctionResult::continue_execution()
    }
}
